src/core/server/elasticsearch/client/cluster_client.ts

```diff
@@ -52,34 +60,70 @@ export interface ICustomClusterClient extends IClusterClient {
 
 /** @internal **/
 export class ClusterClient implements ICustomClusterClient {
-  public readonly asInternalUser: KibanaClient;
+  private readonly config: ElasticsearchClientConfig;
+  private readonly getAuthHeaders: GetAuthHeaders;
+  private readonly setAuthHeaders: SetAuthHeaders;
   private readonly rootScopedClient: KibanaClient;
   private readonly allowListHeaders: string[];
+  private readonly unauthorizedErrorHandler?: UnauthorizedErrorHandler;
+  private readonly getExecutionContext: () => string | undefined;
 
   private isClosed = false;
 
-  constructor(
-    private readonly config: ElasticsearchClientConfig,
-    logger: Logger,
-    type: string,
-    private readonly getAuthHeaders: GetAuthHeaders = noop,
-    getExecutionContext: () => string | undefined = noop
-  ) {
+  public readonly asInternalUser: KibanaClient;
+
+  constructor({
+    config,
+    logger,
+    type,
+    getAuthHeaders = noop,
+    setAuthHeaders = noop,
+    getExecutionContext = noop,
+    unauthorizedErrorHandler,
+  }: {
+    config: ElasticsearchClientConfig;
+    logger: Logger;
+    type: string;
+    getAuthHeaders?: GetAuthHeaders;
+    setAuthHeaders?: SetAuthHeaders;
+    getExecutionContext?: () => string | undefined;
+    unauthorizedErrorHandler?: UnauthorizedErrorHandler;
+  }) {
+    this.config = config;
+    this.getAuthHeaders = getAuthHeaders;
+    this.setAuthHeaders = setAuthHeaders;
+    this.getExecutionContext = getExecutionContext;
+    this.unauthorizedErrorHandler = unauthorizedErrorHandler;
+
     this.asInternalUser = configureClient(config, { logger, type, getExecutionContext });
     this.rootScopedClient = configureClient(config, {
       logger,
       type,
       getExecutionContext,
       scoped: true,
     });
-
     this.allowListHeaders = ['x-opaque-id', ...this.config.requestHeadersWhitelist];
   }
 
   asScoped(request: ScopeableRequest) {
     const scopedHeaders = this.getScopedHeaders(request);
+
+    const internalErrorHandler = this.unauthorizedErrorHandler
```

Comment:
Can we create the handler lazily? We really need it only when `401` happened.

Reply:
We probably can if we want to. Do you have any concern creating is eagerly? It doesn't seems like a costly operation?

Reply:
It's not super expensive, but `asScoped` might be considered as a hot path because Security uses the scoped client(s) on **every** incoming request. But Kibana will need `internalErrorHandler` handler very rarely only when `401` happens after `auth` lifecycle.